### PR TITLE
Ivan Kocherha IM-34

### PR DESCRIPTION
**Pull request by rebte**
Tasks: 
_between_
_compare_
_count_
_day_
_distinct_
_drop_
_duplicate_
_filter_
_words_
_zip_
_intersection_
_invert_

### CODE REVIEW

**Comment by tshemsedinov** [comment]
I''d like to see optimization steps in separate commits

**Comment by rebte** [comment]
need i split it?

Timur Shemsedinov ***@***.***> schrieb am Sa. 13. Jan. 2024
um 14:01:

**Comment by tshemsedinov** [comment]
Sure

File: Tasks/distinct.js [hunk 1]
```diff
@@ -1,18 +1,8 @@
...
-      delete data[w];
-    } else {
-      A.add(a);
-    }
-    w++;
-  });
-  return data.filter
-  (x => typeof x === 'number');
+const distinct = (data) => {
+  return Array.from(new Set(data));
```
Comment (tshemsedinov) [inline Tasks/distinct.js:5]: For example this one: I want to see loop and just after that you can provide alternative without loops

File: Tasks/duplicate.js [hunk 2]
```diff
@@ -1,14 +1,11 @@
...
-    res = [];
-    for (let i = 0; i < N; i++) {
-      res[i] = value;
-    }
-    return res;
+const duplicate = (value, n) => {
+  if (n <= 0) {
+    return [];
   }
+  return Array(n).fill(value);
```
Comment (tshemsedinov) [inline Tasks/duplicate.js:8]: Before this we need loop-based implementation

File: Tasks/filter.js [hunk 3]
```diff
@@ -1,15 +1,8 @@
...
-  for (C of T) {
-    x = T.indexOf(C);
-    if (typeof T[x] !== t) {
-      remove.unshift(x);
-    }
-  }
-  for (x of remove) T.splice(x, 1);
-  return T;
+const filter = (arr, type) => {
+  return arr.filter((el) => typeof el === type);
```
Comment (tshemsedinov) [inline Tasks/filter.js:5]: Ditto

File: package.json [hunk 4]
```diff
@@ -21,6 +21,10 @@
   },
   "homepage": "https://github.com/HowProgrammingWorks/Index",
   "scripts": {
-    "test": "node --test"
+    "test": "node --test",
+    "prettier": "prettier . --write"
```
Comment (tshemsedinov) [inline package.json:25]: You can certainly add prettier but do not commit it to PR

File: .prettierignore [hunk 5]
```diff
@@ -0,0 +1,9 @@
+**/.github
+**/.git
+**/Example
+**/node_modules
+check.js
+package-lock.json
+package.json
+README.md
+.prettierrc.json
```
Comment (tshemsedinov) [inline .prettierignore:9]: All files in repo should have 1 empty line at the end of file

File: package.json (hunk 4 above)
Comment (rebte) [reply package.json:25]: I have a situation where I won’t be able to be at the computer for a week. I can show you the projects later

File: Tasks/distinct.js (hunk 1 above)
Comment (rebte) [reply Tasks/distinct.js:5]: (x => typeof x === 'number') - useless function
w - useless
did better deleting item

File: Tasks/day.js [hunk 6]
```diff
@@ -1,15 +1,15 @@
 // Get day number
+'use strict';
 
-let D = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
+const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
```
Comment (rebte) [inline Tasks/day.js:4]: better variable name

File: Tasks/day.js [hunk 7]
```diff
@@ -1,15 +1,15 @@
...
 
-_parse_day_ = (s) => {
-  let i;
-  for (i = 0; i < D.length; i++) {
-    if (s.startsWith(D[i].toLowerCase())) {
+const parseDay = (str) => {
+  for (let i = 0; i < days.length; i++) {
+    if (str.startsWith(days[i].toLowerCase())) {
```
Comment (rebte) [inline Tasks/day.js:8]: better place for creating loop variable

File: Tasks/duplicate.js (hunk 2 above)
Comment (rebte) [reply Tasks/duplicate.js:8]: `if (n <= 0) {`
`  return [];`
`}`
`const arr = [];`
`for(let i = 0; i < n; i++) {`
`  arr.push(value);`
`}`
`return arr`

File: Tasks/distinct.js (hunk 1 above)
Comment (rebte) [reply Tasks/distinct.js:5]: `const distinct = (data) => {`
`  const k = new Set();`
`  for(let i = 0; i < data.length; i++) {`
`    if(!k.has(data[i])) {`
`      k.add(data[i]);`
`      continue;`
`    }`
`    data.splice(i, 1);`
`    i -= 1;`
`  }`
`  return data;`
`}`

File: Tasks/filter.js (hunk 3 above)
Comment (rebte) [reply Tasks/filter.js:5]: `for(let i = 0; i < arr.length; i++) {`
`  if(typeof arr[i] !== type) {`
`    arr.splice(i, 1);`
`    i -= 1;`
`  }`
`}`
`return arr;`

File: Tasks/count.js [hunk 8]
```diff
@@ -1,17 +1,11 @@
...
-    const value = obj[key];
-    if (typeof value === 'number') sum += value;
-  });
-  ({});
-  return sum;
+const count = (obj) => {
+  return Object.values(obj).reduce(
+    (acc, val) => (typeof val === 'number' ? acc + val : acc),
+    0
+  );
```
Comment (rebte) [inline Tasks/count.js:8]: reduce is good for it

File: Tasks/drop.js [hunk 9]
```diff
@@ -1,24 +1,10 @@
 // Delete listed keys from dictionary
+'use strict';
 
-DroP = (D, ...X) => {
-  T = 100;
```
Comment (rebte) [inline Tasks/drop.js:4]: havent let and useless line

File: Tasks/compare.js [hunk 10]
```diff
@@ -1,18 +1,13 @@
...
-  if (a.join('-') !== b.join('-')) return false;
-  let e = true;
-  for (c of a) {
-    if (first_values[c] === second_values[c]) e = e && true;
-    else {
-      e = e && false;
-    }
+const compare = (first_values, second_values) => {
+  first_values = Object.entries(first_values);
+  second_values = Object.entries(second_values);
```
Comment (tshemsedinov) [inline Tasks/compare.js:6]: Do not mutate incoming arguments

File: Tasks/compare.js [hunk 11]
```diff
@@ -1,18 +1,13 @@
...
-  let a = Object.keys(first_values);
-  let b = Object.keys(second_values);
-  if (a.join('-') !== b.join('-')) return false;
-  let e = true;
-  for (c of a) {
-    if (first_values[c] === second_values[c]) e = e && true;
-    else {
-      e = e && false;
-    }
+const compare = (first_values, second_values) => {
```
Comment (tshemsedinov) [inline Tasks/compare.js:4]: Respect camelCase naming conventions

**Pull request closed without merging**